### PR TITLE
Removes `Shape` tensors

### PR DESCRIPTION
**Pull request by pranavm-nvidia**
In Tripy, we want shapes of tensors to behave like lists. This makes it
easier and more natural to manipulate shapes. However, we previously achieved
this through a `Shape` abstraction that was actually a `Tensor` which behaved
like a list in some, but not all, ways, which made the API very confusing.

To solve this, we remove the `Shape` abstraction and instead represent
shapes of tensors as lists of `DimensionSize`s (which are similar to `ShapeScalar`s).
These lists can be manipulated freely and are then converted back to tensors
by the `convert_to_tensors` decorator, which takes over the functionality of
`convert_shape_inputs`. This also has the advantage that operations on shapes
are not materialized in the computation graph - only operations on the individual
dimensions are.

Also includes various miscellaneous fixes:
- Updates `expand` to disallow `-1`s in prepended dimensions.
- Updates `resize` to use function overloading to accept either output shape or scales.
- Simplifies logic in `flatten`, `repeat`, `unsqueeze`, `full`, and `pad`.
- Greatly simplifies logic in `reshape`.
- Disallows empty lists in `concatenate`, similar to what `stack` does.
- Moves `squeeze` into its own file.
- Fixes to `get_arg_candidate_column_offsets` to work with `skip_num_stack_entries`.
- Updates `Storage` op to copy list inputs so that changes made to a list after it is used to construct a tensor don't affect the created tensor.

### CODE REVIEW

File: tripy/tripy/frontend/trace/ops/base.py [hunk 1]
```diff
@@ -73,80 +72,27 @@ def build(cls, inputs: List["Tensor"], *args, num_outputs=1, **kwargs) -> Union[
         of returning a list of output tensors.
         """
 
-        from tripy.frontend.shape import Shape, ShapeScalar
         from tripy.frontend.tensor import Tensor
+        from tripy.frontend.dimension_size import DimensionSize
+
+        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
+        # For any mixed operations, DimensionSize must be casted up to Tensor.
+        TensorType = DimensionSize if all(isinstance(inp, DimensionSize) for inp in inputs) else Tensor
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/trace/ops/base.py:80]: Still not completely sure of how to implement the logic to determine whether the output should be a DimensionSize or a Tensor. There are two parts to this:
1. What should the decorator convert inputs to? If it always makes Tensors, the operations will never see DimensionSizes and will therefore never be able to indicate the output type accurately. 
2. What type should the op generate?

File: tripy/tripy/frontend/trace/ops/reshape.py [hunk 2]
```diff
@@ -15,40 +15,30 @@
...
 from tripy.frontend.trace.ops import utils as op_utils
 from tripy.frontend.trace.ops.base import BaseTraceOp
+from tripy.types import ShapeLike
 
 
 @dataclass(repr=False)
 class Reshape(BaseTraceOp):
 
     output_rank: int
-    output_len: Optional[int] = None  # only used to help with infer_len for a shape input
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/trace/ops/reshape.py:31]: Need to bring this back for this PR and remove it in a separate PR.

File: tripy/tripy/frontend/utils.py [hunk 3]
```diff
@@ -226,20 +168,49 @@ def add_arg(arg):
...
                 if name in targets and not isinstance(arg, Tensor):
+                    if name in shape_likes:
+                        # NOTE: We cannot use the reshape API here since it would lead to an infinite loop
+                        # when it tries to convert its shape argument via `convert_to_tensors`
+                        dims = [
+                            Reshape.build([dim, Tensor([1])], 1) if isinstance(dim, DimensionSize) else Tensor([dim])
+                            for dim in arg
+                        ]
+                        arg = concatenate(dims, dim=0)
+                        # TODO (pranavm): Remove this WAR for segfault:
```
Comment (pranavm-nvidia) [inline tripy/tripy/frontend/utils.py:179]: Need to fix a segfault in MLIR-TRT before merging this PR.

File: tripy/tripy/frontend/trace/ops/reshape.py (hunk 2 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/reshape.py:31]: Actually this is not needed since `Reshape` can never produce a Shape now.

File: tripy/tests/flat_ir/ops/test_iota.py [hunk 4]
```diff
@@ -49,9 +49,11 @@ def test_mlir(self, flat_ir):
...
             module {
                 func.func @main() -> tensor<?x?xf32> {
-                    %c = stablehlo.constant dense<[2, 3]> : tensor<2xi32>
-                    %0 = stablehlo.dynamic_iota %c, dim = 0 : (tensor<2xi32>) -> tensor<?x?xf32>
-                    return %0 : tensor<?x?xf32>
+                    %c = stablehlo.constant dense<2> : tensor<1xi32>
+                    %c_0 = stablehlo.constant dense<3> : tensor<1xi32>
+                    %0 = stablehlo.concatenate %c, %c_0, dim = 0 : (tensor<1xi32>, tensor<1xi32>) -> tensor<2xi32>
+                    %1 = stablehlo.dynamic_iota %0, dim = 0 : (tensor<2xi32>) -> tensor<?x?xf32>
+                    return %1 : tensor<?x?xf32>
```
Comment (yizhuoz004) [inline tripy/tests/flat_ir/ops/test_iota.py:56]: What is the reason for this change?

File: tripy/tripy/frontend/trace/ops/slice.py [hunk 5]
```diff
@@ -42,46 +40,8 @@ def infer_rank(self):
     def infer_len(self):
         # Only infer if we have concrete values to use. Note that the result is only a Shape if these are *slices*,
         # not single indices, so a slice is the only case that needs to be considered
-        if self.shape_slice is not None:
```
[the remark this answers is not in the document]
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/slice.py:45]: Since shapes are just lists now, slicing shapes does not trigger this code path.

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (yizhuoz004) [reply tripy/tripy/frontend/trace/ops/base.py:80]: Shall we only support a few ops to return `DimensionSize`? E.g. only binary and unary elementwise. I did not see other use cases. To return `DimensionSize`, the op must follow quite strict rules, e.g. I/O rank must be 0, dtype must be int32, it may be easier to just white-list the ops and let them override some function to provide the type information.

**Comment by yizhuoz004** [comment]
Since tensor.shape is just a list now, how do we support dynamic slicing? for example
```
d0 = a.shape[0] # d0 is a DimensionSize
b.shape[d0]
```

File: tripy/docs/post0_developer_guides/how-to-add-new-ops.md [hunk 6]
```diff
@@ -200,11 +196,10 @@ import tripy.frontend.utils as frontend_utils
 # If we needed to provide any special autodoc options, we could use the `autodoc_options` parameter.
 @export.public_api(document_under="tensor_operations")
 
-# The `convert_shape_inputs` decorator converts the specified function arguments into `tripy.Shape`s,
-# which would allow for using Python numbers and sequences. The `convert_to_tensors` decorator more generally converts
-# function arguments into Tripy tensors and is also commonly used in the codebase.
-@frontend_utils.convert_shape_inputs(["shape"])
-def theta(shape: Tuple[int], dim: int = 0, dtype: datatype.dtype = datatype.float32) -> "tripy.Tensor":
+# The `convert_to_tensors` decorator converts automatically converts compatible arguments,
+# like `TensorLike` or `ShapeLike`s into tensors.
```
Comment (slyubomirsky) [inline tripy/docs/post0_developer_guides/how-to-add-new-ops.md:200]:
```suggestion
# The `convert_to_tensors` decorator automatically converts compatible arguments,
# like `TensorLike` or `ShapeLike`s, into tensors.
```

**Comment by pranavm-nvidia** [comment]
> Since tensor.shape is just a list now, how do we support dynamic slicing? for example
> 
> ```
> d0 = a.shape[0] # d0 is a DimensionSize
> b.shape[d0]
> ```


If we support `b.shape[d0]`, then it seems natural to support `b.shape[:d0]` also . This would break our frontend rank/len inference because we would have to know the value of the shape elements instead of just the number of elements. Practically speaking, I can't really imagine a case where you'd need to index a shape based on the value of another shape. I expect that you would be able to transform the shape value (i.e. `d0`) directly in such cases.

File: tripy/tests/flat_ir/ops/test_iota.py (hunk 4 above)
Comment (slyubomirsky) [reply tripy/tests/flat_ir/ops/test_iota.py:56]: Perhaps a consequence of having lists of dimensions rather than single shape tensors?

File: tripy/tripy/frontend/dimension_size.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                raise_error(
+                    f"Scalar shape tensors must be of rank 0, but input tensor is rank {data.rank}", details=[data]
+                )
+            if data.dtype is not None and data.dtype != int32:
+                raise_error(
+                    f"Scalar shape tensor must have int32 member, but input tensor has data type {data.dtype}",
+                    details=[data],
+                )
+
+            # the shape of data should correspond to the given rank
```
Comment (slyubomirsky) [inline tripy/tripy/frontend/dimension_size.py:54]:
```suggestion
```
I don't think the comment refers to anything now.

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py:80]: I agree. I'm stuck on the mechanism for whitelisting though. What we would need is for the frontend APIs to somehow have knowledge of the original input types before they are converted to tensors. With that in place, we could just have `convert_to_tensors` always convert to the `Tensor` base class - since `Tensor`/`DimensionSize` are just frontend constructs, we can actually consume `Tensor`s and produce `DimensionSize`. 

One approach is to attach the original type as a private field to the tensor produced by `convert_to_tensors`:
```py
@convert_to_tensors()
def func(a: TensorLike, b: TensorLike):
    return_type = Tensor
    if a._original_type is DimensionSize and b._original_type is DimensionSize:
        return_type = DimensionSize
```
But I'm wondering if there is a better/cleaner approach than this.

File: tripy/tests/flat_ir/ops/test_iota.py (hunk 4 above)
Comment (jhalakpatel) [reply tripy/tests/flat_ir/ops/test_iota.py:56]: I wonder if the same IR bloat will be present once we move to the tensorrt dialect.

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (slyubomirsky) [reply tripy/tripy/frontend/trace/ops/base.py:80]: Is there any reason not to convert to `DimensionSize` if the criteria are met?

File: tripy/tests/flat_ir/ops/test_iota.py (hunk 4 above)
Comment (yizhuoz004) [reply tripy/tests/flat_ir/ops/test_iota.py:56]: I did not understand the difference, the original one should just work.

**Comment by yizhuoz004** [comment]
> > Since tensor.shape is just a list now, how do we support dynamic slicing? for example
> > ```
> > d0 = a.shape[0] # d0 is a DimensionSize
> > b.shape[d0]
> > ```
> 
> If we support `b.shape[d0]`, then it seems natural to support `b.shape[:d0]` also . This would break our frontend rank/len inference because we would have to know the value of the shape elements instead of just the number of elements. Practically speaking, I can't really imagine a case where you'd need to index a shape based on the value of another shape. I expect that you would be able to transform the shape value (i.e. `d0`) directly in such cases.

I agree. We still support dynamic slicing if the input is transformed into a tensor.

File: tripy/tests/test_function_registry.py [hunk 8]
```diff
@@ -474,7 +474,7 @@ def func(n: Sequence[Union[int, float]]) -> int:
     "typ, expected",
     [
         (tp.types.TensorLike, "Union[tripy.Tensor, numbers.Number]"),
-        (tp.types.ShapeLike, "Union[tripy.Shape, Sequence[Union[int, tripy.ShapeScalar]]]"),
+        (tp.types.ShapeLike, "Sequence[Union[int, tripy.DimensionSize]]"),
```
Comment (jhalakpatel) [inline tripy/tests/test_function_registry.py:477]: shouldn't `tripy.DimensionsSize` be an instance of `TensorLike` since it inherits from the `Tensor` class?

File: tripy/tripy/frontend/trace/ops/slice.py [hunk 9]
```diff
@@ -42,46 +40,8 @@ def infer_rank(self):
     def infer_len(self):
         # Only infer if we have concrete values to use. Note that the result is only a Shape if these are *slices*,
         # not single indices, so a slice is the only case that needs to be considered
```
Comment (slyubomirsky) [inline tripy/tripy/frontend/trace/ops/slice.py:42]:
```suggestion
```
Since this gets rid of `Shape` as a class, the comment is no longer applicable either.

File: tripy/tests/flat_ir/ops/test_iota.py (hunk 4 above)
Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_iota.py:56]: Yes, it's a consequence of having lists of dimensions. `Shape([2, 3])` will materialize the two element tensor, whereas now we are doing a `concat(DimensionSize([2]), DimensionSize([3]))`.

File: tripy/tripy/frontend/utils.py [hunk 10]
```diff
@@ -570,30 +386,33 @@ def topological_sort(ops: List[Union["BaseTraceOp", BaseFlatIROp]]) -> List[Unio
...
-                    ],
-                )
-            return new_dim
-
-        args, kwargs = utils.modify_arg("dim", process_dim, func, *args, **kwargs)
-        return func(*args, **kwargs)
-
-    return wrapper
+def process_dim():
+    def process_dim_impl(func):
```
Comment (slyubomirsky) [inline tripy/tripy/frontend/utils.py:390]: Was there a specific reason to add this step?

File: tripy/tripy/frontend/trace/ops/reshape.py [hunk 11]
```diff
@@ -58,26 +48,19 @@ def infer_rank(self):
         else:
             self.outputs[0].rank = self.output_rank
 
-    @frontend_utils.make_function
+    # @frontend_utils.make_function
```
[the remark this answers is not in the document]
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/reshape.py:51]: I don't think so since it's just a single FlatIR op. Will remove it.

File: tripy/tests/test_function_registry.py (hunk 8 above)
Comment (pranavm-nvidia) [reply tripy/tests/test_function_registry.py:477]: Yes, it is, since `TensorLike` includes `Tensor` (and therefore all children of `Tensor`).

File: tripy/tripy/frontend/dimension_size.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            super().__init__(data=None, dtype=int32, name=name, device=data.device)
+            # share the underlying data
+            self.trace_tensor = data.trace_tensor
+            self.stack_info = data.stack_info
+        else:
+            # NOTE: We do not use isinstance here because bool is a subclass of int.
```
[the remark this answers is not in the document]
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/dimension_size.py:60]: If we did then you could have `DimensionSize`s in the frontend with boolean values. Then we have to decide what it would mean to do `reshape(x, [True, False, True, False])` :).

File: tripy/tripy/utils/ast.py [hunk 13]
```diff
@@ -141,10 +137,6 @@ def index_into_expr(node: ast.expr, index: int) -> ast.expr:
             else:
                 arg_node = index_into_expr(node.slice, index)
 
-        # if the resulting arg node is a list or tuple literal, we might be able to narrow it down with a list index
-        if list_index is not None and isinstance(arg_node, (ast.List, ast.Tuple)) and list_index < len(arg_node.elts):
-            arg_node = arg_node.elts[list_index]
```
Comment (slyubomirsky) [inline tripy/tripy/utils/ast.py:146]: What was the reason for getting rid of this? Did it cause problems somewhere?

Comment (pranavm-nvidia) [reply tripy/tripy/utils/ast.py:146]: It was dead code

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py:80]: Since `DimensionSize` -> `Tensor` is a one-way conversion, we should try to keep things as `DimensionSize`s where possible. The problem is that the logic that I've currently implemented isn't completely right - not all Python integers actually represent dimension sizes and not all ops produce dimension sizes even if all their inputs are dimension sizes.

Comment (yizhuoz004) [reply tripy/tripy/frontend/trace/ops/base.py:80]: Is it possible to wrap another decorator to decide the return type? e.g.

```py
@infer_return_type
@convert_to_tensors()
def func(...)
```

File: tripy/tripy/utils/ast.py (hunk 13 above)
Comment (slyubomirsky) [reply tripy/tripy/utils/ast.py:146]: Ah, I see, the code above handles it too.

File: tripy/tripy/frontend/dimension_size.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (slyubomirsky) [reply tripy/tripy/frontend/dimension_size.py:59]: Good point, probably not a case where treating Bools as numbers would be advisable.

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (slyubomirsky) [reply tripy/tripy/frontend/trace/ops/base.py:80]: > not all Python integers actually represent dimension sizes

We could present `DimensionSize` as meaning that the value _could_ be used to represent a dimension size, but maybe that's too clever by half. I think using a method or perhaps another decorator to indicate a policy for wrapping might be a sensible approach if isn't a way to avoid the issue.

Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py:80]: Yes, an outer decorator is a good idea. Will update the PR accordingly.

File: tripy/tripy/frontend/utils.py [hunk 14]
```diff
@@ -19,43 +19,40 @@
...
+    from tripy.common.datatype import int32
+    from tripy.frontend.dimension_size import DimensionSize
     from tripy.frontend.tensor import Tensor
-    from tripy.frontend.trace.ops.cast import cast
+    from tripy.frontend.trace.ops.concatenate import concatenate
+    from tripy.frontend.trace.ops.reshape import Reshape
+
+    if not arg:
+        return Tensor([], dtype=int32)
+
```
Comment (yizhuoz004) [inline tripy/tripy/frontend/utils.py:39]: Might worth adding a simple optimization:
```py
if not arg or all(isinstance(e, int) for e in arg):
    return Tensor(arg, dtype=int32)
```

Comment (pranavm-nvidia) [reply tripy/tripy/frontend/utils.py:39]: @yizhuoz004 Thanks, I implemented a generalized version of this approach which will accumulate contiguous sequences of integers instead of creating individual tensors for each.

File: tripy/tests/flat_ir/ops/test_iota.py (hunk 4 above)
Comment (pranavm-nvidia) [reply tripy/tests/flat_ir/ops/test_iota.py]: This is resolved by the fix for: https://github.com/NVIDIA/TensorRT-Incubator/pull/306#discussion_r1821340540
The IR should now be the same as before.

**Review by yizhuoz004 — APPROVED** [review]

File: tripy/tripy/frontend/trace/ops/base.py (hunk 1 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py]: Updated this as we discussed so that we always try casting to DimensionSize where possible.

File: tripy/tripy/frontend/trace/ops/base.py [hunk 15]
```diff
@@ -72,21 +73,30 @@ def build(cls, inputs: List["Tensor"], *args, num_outputs=1, **kwargs) -> Union[
...
         inp_trace_tensors = [inp.trace_tensor for inp in inputs]
         out_trace_tensors = [out.trace_tensor for out in outputs]
         cls.build_internal(inp_trace_tensors, out_trace_tensors, *args, **kwargs)
 
+        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
+        # For any mixed operations, DimensionSize must be casted up to Tensor.
+        all_inputs_are_dimension_size = all(isinstance(inp, DimensionSize) for inp in inputs)
+        for index, out in enumerate(outputs):
+            if always_cast_to_dimension_size or (
+                all_inputs_are_dimension_size and out.dtype == int32 and out.shape == []
```
Comment (yizhuoz004) [inline tripy/tripy/frontend/trace/ops/base.py:93]: Will `out.shape` trigger eval here?

File: tripy/tripy/frontend/trace/ops/base.py [hunk 16]
```diff
@@ -72,21 +73,30 @@ def build(cls, inputs: List["Tensor"], *args, num_outputs=1, **kwargs) -> Union[
...
         # the Tensor constructor!
-        outputs = [TensorType(None) for _ in range(num_outputs)]
+        outputs = [Tensor(None) for _ in range(num_outputs)]
 
         inp_trace_tensors = [inp.trace_tensor for inp in inputs]
         out_trace_tensors = [out.trace_tensor for out in outputs]
         cls.build_internal(inp_trace_tensors, out_trace_tensors, *args, **kwargs)
 
+        # Operations that operate on only DimensionSize inputs will always yield a DimensionSize.
+        # For any mixed operations, DimensionSize must be casted up to Tensor.
```
Comment (yizhuoz004) [inline tripy/tripy/frontend/trace/ops/base.py:89]: Shall we add simple unit tests to guard this logic?

**Review by parthchadha — APPROVED** [review]

File: tripy/tripy/frontend/trace/ops/base.py (hunk 15 above)
Comment (slyubomirsky) [reply tripy/tripy/frontend/trace/ops/base.py:93]: I think not since these would be trace tensors.

**Review by slyubomirsky — APPROVED** [review]

File: tripy/tripy/frontend/trace/ops/base.py (hunk 16 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py:89]: @yizhuoz004 I did add a test under `tripy/tests/frontend/trace/ops/test_binary_elementwise.py`. I could add a separate test with a mock operation, but I don't think it provides any additional coverage.

File: tripy/tripy/frontend/trace/ops/base.py (hunk 15 above)
Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py:93]: Oh good catch, yes it would. I meant to do `out.trace_tensor.shape`

Comment (pranavm-nvidia) [reply tripy/tripy/frontend/trace/ops/base.py]: Fixed.
